File: javascript/018/main.js
/*
 * ProjectEuler 018
 * JavaScript / main.js
 * copyright (c) 2015 Susisu
 *
 * Node.js v4.2.1
 * `node main.js`
 */

"use strict";

var fs = require("fs");

function main() {
    fs.readFile("./numbers.txt", { "encoding": "UTF-8" }, function (err, res) {
        if (err) {
            console.log(err);
            return;
        }
        var nums = res.split("\n")
            .filter(function (line) { return line.length > 0; })
            .map(function (line) {
                return line.split(" ")
                    .map(function (num) { return parseInt(num); });
            });
        for (var i = 1; i < nums.length; i++) {
            var row = nums[i];
            for (var j = 0; j < row.length; j++) {
                if (j == 0) {
                    row[j] += nums[i - 1][j];
                }
                else if (j < row.length - 1) {
                    row[j] += Math.max(nums[i - 1][j], nums[i - 1][j - 1]);
                }
                else {
                    row[j] += nums[i - 1][j - 1];
                }
            }
        }
        var max = Math.max.apply(undefined, nums[nums.length - 1]);
        console.log(max);
    });
}

main();

File: ocaml/018/main.ml
(*
    ProjectEuler 018
    OCaml / main.ml
    copyright (c) 2015 Susisu

    OCaml 4.02.3
    `ocamlfind ocamlopt -o bin/main -linkpkg -package core -thread nums.cmxa main.ml`
*)

open Core

let descend xs ys =
    let xs_ = xs @ [0] in
    let _xs = 0 :: xs in
    let xs' = List.map2 max xs_ _xs in
    List.map2 (+) xs' ys

let line_int_stream_of_channel ic =
    Stream.from (fun _ ->
        try
            let line = input_line ic in
            let nums = List.map int_of_string (Core_string.split line ' ') in
            Some nums
        with End_of_file -> None
    )

let read_numbers_line ic =
    let line = input_line ic in
    let nums = Core_string.split line ' ' in
    List.map (int_of_string) nums

let main _ =
    let ic = open_in "./numbers.txt" in
    try
        let st = line_int_stream_of_channel ic in
        let ns = Stream.npeek 15 st in
        let b  = List.fold_left descend [] ns in
        print_int (List.fold_left max 0 b);
        print_newline ();
        close_in ic
    with
    | err ->
        close_in_noerr ic;
        raise err

let _ = main ()
